### PR TITLE
Jim/lineage3

### PR DESCRIPTION
**Pull request by jimduff-usds**
This PR does further work on the audit trail tables, in particular, preparing for tracking of item-level data.

## Changes
- Added audit trail tracking to DownloadFunction.
- Addition of example lineage/history queries
- Added lots more database data to the `DatabaseAccess.Header` container, which will get renamed eventually.   This data will be used to implement item-level tracking.   
- Begin to reduce dependence on the Task and TaskSource tables.  Further work to come.
- The audit trail not only tracks internal reports, it also adds a row to report_file table for every exterally delivered report, by redox, sftp, or download, for example.
- All authenticated calls to our APIs are now tracked in the audit trail `action` table, even including things like option=CheckConnections
- Slight increase in security of download:   Not only do you have to pass Authentication, but your organization must also exist in the set of organizations in our metadata.
- fixed race condition where Merge function started before the data it needed was committed to the Task table.
- Now sending Header to the EmailTransport.buildContent.  Its available,to now add a lot more useful detail to the email.
- Filenames used by sftp send are now identical to the download site filesnames.

## Checklist
- [X] Tested locally?
- [ ] Added new dependencies?
- [X] Updated the docs?
- [ ] Added a test?
- [X] Did you check for sensitive data, and remove any?
- [ ] Are additional approvals needed for this change?
- [ ] Are there potential vulnerabilities or licensing issues with any new dependencies introduced?

## Fixes 

- #433 Moving Canyons to prod.
- #147 Report Audit Trail

## To Be Done

- #426 Item level audit trail
- Identified issue where very large merges (20 files) got split between two Batch function runs.

### CODE REVIEW

File: prime-router/src/main/kotlin/azure/DownloadFunction.kt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package gov.cdc.prime.router.azure
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/azure/DownloadFunction.kt:1]: @mry-usds please review my changes to DownloadFunction

File: prime-router/src/main/kotlin/azure/ActionHistory.kt [hunk 2]
```diff
@@ -133,6 +147,11 @@ class ActionHistory {
         trackActionResult(httpResponseMessage.status.toString() + ":\n" + httpResponseMessage.body.toString())
     }
 
+    fun trackActionRequestResponse(request: HttpRequestMessage<String?>, response: HttpResponseMessage) {
+        trackActionParams(request)
+        trackActionResult(response)
+    }
+
```
[the remark this answers is not in the document]
Comment (mry-usds) [reply prime-router/src/main/kotlin/azure/ActionHistory.kt:154]: Good

File: prime-router/src/main/kotlin/azure/DatabaseAccess.kt [hunk 3]
```diff
@@ -41,10 +45,35 @@ class DatabaseAccess(private val create: DSLContext) {
     constructor(dataSource: DataSource) : this(DSL.using(dataSource, SQLDialect.POSTGRES))
     constructor(connection: Connection) : this(DSL.using(connection, SQLDialect.POSTGRES))
 
-    data class Header(val task: Task, val sources: List<TaskSource>)
+    class Header(
```
[the remark this answers is not in the document]
Comment (mry-usds) [reply prime-router/src/main/kotlin/azure/DatabaseAccess.kt:48]: Risk is that header (which used to be lightweight) - now becomes a heavy object with content...consider deferring loading of content until it is requested...(i.e. keep it light weight unless it needs to be heavy)

File: prime-router/src/main/kotlin/azure/DownloadFunction.kt [hunk 4]
```diff
@@ -37,22 +43,28 @@ class DownloadFunction {
...
             methods = [HttpMethod.GET],
             authLevel = AuthorizationLevel.ANONYMOUS
-        ) request: HttpRequestMessage<String?>
+        ) request: HttpRequestMessage<String?>,
+        context: ExecutionContext,
     ): HttpResponseMessage {
-        var authenticated = checkAuthenticated(request)
-        if (authenticated.first) {
+        var authClaims = checkAuthenticated(request, context)
+        if (authClaims != null) {
```
Comment (mry-usds) [inline prime-router/src/main/kotlin/azure/DownloadFunction.kt:62]: cleaner - easier to read

File: prime-router/src/main/kotlin/azure/DownloadFunction.kt [hunk 5]
```diff
@@ -178,13 +220,16 @@ class DownloadFunction {
...
+     * returns null if not authorized, otherwise returns a set of claims.
+     */
+    private fun checkAuthenticated(request: HttpRequestMessage<String?>, context: ExecutionContext): AuthClaims? {
         var userName = ""
         var orgName = ""
 
         val cookies = request.headers["cookie"] ?: ""
 
-        System.out.println(cookies)
+        context.logger.info(cookies)
```
Comment (mry-usds) [inline prime-router/src/main/kotlin/azure/DownloadFunction.kt:232]: should remove entirely.

**Review by mry-usds — APPROVED** [review]

**Comment by jimduff-usds** [comment]
@mry-usds your point about Header getting too heavyweight is a good one.  I'll look at lazy instantiation in the next round.

File: prime-router/docs/proposals/0003-lineage/additional-notes.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+Every report in the database must be generated by an action.
+
+The initial `receive` action does quite a lot:  it generates both the incoming file, and the transformed reports for each future receiver as well.
+
+I considered naming this table REPORT_HIST instead of REPORT_FILE, and making it immutable.  I can think of a couple reasons not to do this:
+1. Bad data.   If the system generates erroneous data, you need a way to mark that data as bad.
+2. The WIPE action - I feel like if you wipe a report from the system, the REPORT_FILE table should be updated to reflect that the data no longer exists.
+One option would be to implement triggers that prevent updates except for a status field.  Note:  the report_file table currently has no status field.
+
```
Comment (RickHawesUSDS) [inline prime-router/docs/proposals/0003-lineage/additional-notes.md:21]: +1 on the idea of REPORT_HISTORY (spell out words please).

File: prime-router/src/main/resources/db/migration/V4__lineage_stored_procs.sql [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+ */
+
+/* 
+ * This SQL makes some tweaks to report_file table,
+ * and it adds basic stored procedures to do recursive lineage queries.
+ */
+
+ALTER TABLE report_file DROP CONSTRAINT report_file_external_name_key;
+ALTER TABLE report_file ADD COLUMN downloaded_by VARCHAR(63);  -- null ok.
+
```
[the remark this answers is not in the document]
Comment (RickHawesUSDS) [reply prime-router/src/main/resources/db/migration/V4__lineage_stored_procs.sql:18]: +1

File: prime-router/src/main/kotlin/azure/DownloadFunction.kt (hunk 5 above)
Comment (RickHawesUSDS) [reply prime-router/src/main/kotlin/azure/DownloadFunction.kt:232]: Yes. Please remove.

File: prime-router/src/main/kotlin/azure/DownloadFunction.kt [hunk 8]
```diff
@@ -139,27 +158,50 @@ class DownloadFunction {
...
+                WorkflowEngine().recordAction(actionHistory)
+                return response
             }
         } catch (ex: Exception) {
-            response = request.createResponseBuilder(HttpStatus.NOT_FOUND).build()
+            context.logger.log(Level.WARNING, "Exception during download of $requestedFile", ex)
+            response = request.createResponseBuilder(HttpStatus.NOT_FOUND)
+                .body("File $requestedFile not found")
+                .header("Content-Type", "text/html")
+                .build()
```
[the remark this answers is not in the document]
Comment (RickHawesUSDS) [reply prime-router/src/main/kotlin/azure/DownloadFunction.kt:203]: +1 (but a design is called for).

**Review by RickHawesUSDS — COMMENTED** [review]
Looks really good. Lot's of improvements. Thanks.

**Pull request merged**